### PR TITLE
Port line-dasharray fix from gl-js

### PR DESCRIPTION
**Pull request by karimnaaji**
Port of bug-fix on line-dasharray from gl-js.

For background, refer:
- https://github.com/mapbox/mapbox-gl-js/issues/9213
- https://github.com/mapbox/mapbox-gl-js/pull/9246

Closes https://github.com/mapbox/mapbox-gl-native/issues/16181

## Launch Checklist

- [x] Enable render test https://github.com/mapbox/mapbox-gl-native/blob/master/metrics/ignores/platform-all.json#L135
 - [x] apply `needs changelog` label if a changelog is needed (remove label when added)

### CODE REVIEW

File: src/mbgl/geometry/line_atlas.cpp [hunk 1]
```diff
@@ -20,82 +19,142 @@ size_t getDashPatternHash(const std::vector<float>& dasharray, const LinePattern
...
+}
+
+void addRoundDash(
+    const std::vector<DashRange>& ranges, const int32_t yOffset, float stretch, const int n, AlphaImage& image) {
+    const float halfStretch = stretch * 0.5f;
 
     for (int y = -n; y <= n; y++) {
         int row = yOffset + n + y;
         int index = image.size.width * row;
+        int currIndex = 0;
```
Comment (alexshalamov) [inline src/mbgl/geometry/line_atlas.cpp]: can wraparound to negative value, use unsigned or std::size_t?

File: src/mbgl/geometry/line_atlas.cpp [hunk 2]
```diff
@@ -20,82 +19,142 @@ size_t getDashPatternHash(const std::vector<float>& dasharray, const LinePattern
...
+
+    DashRange& first = ranges.front();
+    DashRange& last = ranges.back();
+    if (first.isDash == last.isDash) {
+        first.left = last.left - image.size.width;
+        last.right = first.right + image.size.width;
+    }
+
+    int index = image.size.width * yOffset;
+    int currIndex = 0;
```
Comment (alexshalamov) [inline src/mbgl/geometry/line_atlas.cpp]: bot can overflow

**Comment by karimnaaji** [comment]
Thanks for the info the baselines @alexshalamov ! It's updated, ready for another review.

**Comment by alexshalamov** [comment]
@karimnaaji rebased & meged.

**Pull request merged**